src/effects/RealtimeEffectManager.cpp

```diff
@@ -274,6 +276,55 @@ void RealtimeEffectManager::VisitAll(StateVisitor func)
       RealtimeEffectList::Get(*leader).Visit(func);
 }
 
+RealtimeEffectState *
+RealtimeEffectManager::AddState(Track *pTrack, const PluginID & id)
+{
+   auto pLeader = pTrack ? *TrackList::Channels(pTrack).begin() : nullptr;
+   RealtimeEffectList &states = pLeader
+      ? RealtimeEffectList::Get(*pLeader)
+      : RealtimeEffectList::Get(mProject);
+
+   SuspensionScope scope{ &mProject };
+   // Protect...
+   std::lock_guard<std::mutex> guard(mLock);
```

Comment:
mLock here will be locked/unlocked three times:

1) locked and then released as a result of the call to ::Suspend() in the suspension scope constuctor
2) locked by `guard`, then released by its destructor when return is reached
3) locked and then released as a result of a call to Resume() in the supension scope destructor

There is potentially unprotected time between a release and a subsequent lock. I wonder if all of these operations should better be under only one lock. One solution could be to have non-locking versions of Suspend() and Release()

Reply:
Very good.  I noticed the same too.

This is in a function that runs in the main thread.  The functions that can run in another thread are `ProcessStart`, `Process`, `ProcessEnd`.  These are the calls that really need mutual exclusion from the main thread functions, so that the contents of the `RealtimeEffectList`s are stable for that thread.

Each of those functions checks first whether processing is suspended, and they do nothing if so.

I don't think there is any incorrectness in the avoidance of races, but there is probably needless inefficiency, and a lock-free alternative should be sought.

That is not yet in the intended scope of any of these pull requests.